### PR TITLE
Differentiate between inactive/end-dated

### PR DESCRIPTION
**Pull request by steve-gregory**
### End-Dated providers:
End-dated providers are not visible within the GUI/API (Archived)
To include end-dated providers in API responses use '?archived=True'
End-dated providers will throw ProviderNotActive exception.

### Inactive providers:
Inactive providers are visible with the GUI/API
Inactive providers _are not available for instance launch_ within the GUI/API
Inactive providers can still have drivers created.

## Checklist before merging Pull Requests
- [ ] Reviewed and approved by at least one other contributor.
- [ ] If necessary, include a snippet in CHANGELOG.md

### CODE REVIEW

File: core/models/provider.py [hunk 1]
```diff
@@ -203,7 +203,23 @@ def get_platform_name(self):
     def get_type_name(self):
         return self.type.name
 
+    def is_current(self):
+        """
+        Is current when:
+        - No end date present
+        - End date is earlier than 'now'
```
Comment (cdosborn) [inline core/models/provider.py]: Do you mean end date is later than now?

File: core/models/provider.py [hunk 2]
```diff
@@ -203,7 +203,23 @@ def get_platform_name(self):
...
+        """
+        if not self.end_date:
+            return True
+        now = timezone.now()
+        return self.end_date < now
```
Comment (cdosborn) [inline core/models/provider.py]: This logic can be cleaned up, return a single boolean expression.

I think this should be:
```
return not self.end_date || self.end_date > now
```

File: core/models/provider.py [hunk 3]
```diff
@@ -203,7 +203,23 @@ def get_platform_name(self):
...
+
     def is_active(self):
+        """
+        Is active when:
+        - No end date present OR end date is earlier than now
+        - active' == True
+        """
```
Comment (cdosborn) [inline core/models/provider.py]: This needs to be fixed, its possible that `this.active == True` and `this.is_active() == False`.

File: core/query.py [hunk 4]
```diff
@@ -187,7 +184,7 @@ def _in_range():
 
     if not now_time:
         now_time = timezone.now()
-    return _in_range() & _active_provider()
+    return _in_range() & current_provider()
```
Comment (cdosborn) [inline core/query.py]: Use `&&`, `&` does not short circuit
https://stackoverflow.com/a/3845032/1213041

**Comment by cdosborn** [comment]
What is the use case for this pr?

File: core/query.py (hunk 4 above)
Comment (cdosborn) [inline core/query.py]: We can ignore this since it is unrelated.

Comment (steve-gregory) [inline core/query.py]: I believe this is actually overuse of the `&` operator to create a `query & query` as the result returned.

File: core/models/provider.py (hunk 3 above)
Comment (steve-gregory) [inline core/models/provider.py]: an end-dated provider will always be inactive.. 😄

File: core/models/provider.py (hunk 1 above)
Comment (steve-gregory) [inline core/models/provider.py]: yes 👍

File: core/models/provider.py (hunk 3 above)
Comment (steve-gregory) [inline core/models/provider.py]: The documentation of the function has been updated accordingly 👍

File: core/query.py (hunk 4 above)
Comment (cdosborn) [inline core/query.py]: oh hell i'm so wrong

**Comment by julianpistorius** [comment]
Is this PR supposed to be go in to the release?

**Comment by steve-gregory** [comment]
:shipit: because this needs to be part of the `v27` release to include our `deprecation strategy` for old cloud providers.
